### PR TITLE
CAKE-4741 - Trivia Quizzes MW integration

### PR DESCRIPTION
**Pull request by jamieshepherd**
## Description
- Preview our upcoming new content type "Trivia Quizzes" in Mediawiki articles on enabled alpha communities (currently only Game of Thrones)
- Creates a new wikia/ extension `TriviaQuizzes`, which under certain conditions replaces the "Recent Wiki Activity" Oasis module, with the Trivia Quizzes right rail embed
- 2 new "wgVariables", to enable the module at the community level, and the page level
- Commits the `content-types-consumption` distribution to code, this needs to be re-built on updates
- Check for GDPR tracking

## Please note
Code living in `LatestActivityController` and `LatestActivity_Index.php` is **NOT INTENDED TO BE PERMANENT.** We have separate clean-up tickets if we prove out the experiment to make Trivia Quizzes a first class extension, instead of the piggy-backing in the right rail it's currently doing. As a result, the code is meant to do the job in the "fastest possible path" with minimal disruption.

Clean-up ticket: https://wikia-inc.atlassian.net/browse/CAKE-4746 (Please add longer-term suggestions here if necessary).

## Work in progress
The front-end module is still being finished - and will be thoroughly QA'd and design reviewed next week. This PR should represent only the implementation of this module. Separate QA and design updates will come in the form of an updated dist module.

## Test
Check it out on https://gameofthrones.jshepherd.fandom-dev.us/wiki/Jon_Snow - and for a page it should **NOT** be loaded on https://gameofthrones.jshepherd.fandom-dev.us/wiki/Test_Page

@Wikia/cake 🍰 
@Wikia/platform-team

### CODE REVIEW

File: extensions/wikia/TriviaQuizzes/TriviaQuizzes.setup.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
+<?php
+$wgExtensionCredits['other'][] = [
+	'name' => 'Trivia Quizzes',
```
Comment (mklucsarits) [inline extensions/wikia/TriviaQuizzes/TriviaQuizzes.setup.php:3]: Something is funky with the indentation in this file. Can you make sure you're using Tab indentation for all the PHP files?

**Comment by JustLeafy** [comment]
Is there a particular reason why we are replacing the Recent Wiki Activity module with a Trivia Quizzes module? I think it would be better to make the rail available on Special:WikiActivity and on Feeds. Not only it wouldn't be a replacement, but it would keep some practical use for the original module; i.e. seeing the most recent edits on the wiki without the need to visit the page itself. Just my two cents on this plan.

File: includes/wikia/VariablesBase.php [hunk 2]
```diff
@@ -8964,3 +8964,17 @@
  * @var bool
  */
 $wgEnableQualtricsSiteInterceptExt = false;
+
+/**
+ * Enables Quizzes extension on select communities
+ * @see CAKE-4585
+ * @var bool
+ */
+$wgEnableTriviaQuizzesAlpha = false;
```
Comment (mklucsarits) [inline includes/wikia/VariablesBase.php]: Let's name this like `$wgEnableTriviaQuizzesExt` so that (1) we know it's the on/off switch for the extension and (2) it follows the naming convention of on/off variables for other MW extensions.

Comment (JustLeafy) [reply includes/wikia/VariablesBase.php]: Alright, that's fair enough.

**Comment by jamieshepherd** [comment]
@JustLeafy - The extension is a content types experiment we're going to launch over the next few weeks - but we'll be restricting it to very select communities (and article pages) to get feedback. The product requirement is that its entry point lives in the right rail, and on these specific few pages will replace the "Recent Wiki Activity" module.

If the experiment proves out to be successful, we expect it will live as its own module outright, and the code in LatestActivity will be removed. This is a temporary "fastest-path" solution. Appreciate the concern.

File: extensions/wikia/TriviaQuizzes/TriviaQuizzes.setup.php (hunk 1 above)
Comment (jamieshepherd) [reply extensions/wikia/TriviaQuizzes/TriviaQuizzes.setup.php:3]: Ran through stylize.php 0e890bbd5de6b30e08aaa907fe0da845389ad320

File: extensions/wikia/TriviaQuizzes/package.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (mklucsarits) [inline extensions/wikia/TriviaQuizzes/package.json:1]: Why is this file needed?

File: extensions/wikia/TriviaQuizzes/yarn.lock [hunk 4]
```diff
@@ -0,0 +1,325 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (mklucsarits) [inline extensions/wikia/TriviaQuizzes/yarn.lock]: Why is this file needed?

File: extensions/wikia/TriviaQuizzes/package.json (hunk 3 above)
Comment (jamieshepherd) [reply extensions/wikia/TriviaQuizzes/package.json:1]: To declare the extension's dependencies (`npm install` or `npm update` will fetch the latest version of content-types-consumption).

File: extensions/wikia/TriviaQuizzes/TriviaQuizzesHooks.class.php [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+class TriviaQuizzesHooks {
+    public static function onBeforePageDisplay( OutputPage $out ) {
```
Comment (Grunny) [inline extensions/wikia/TriviaQuizzes/TriviaQuizzesHooks.class.php]: Minor: All PHP code should be using tabs, not spaces.

File: extensions/wikia/TriviaQuizzes/TriviaQuizzesHooks.class.php [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+        global $wgTitle, $wgEnableTriviaQuizzesExt, $wgTriviaQuizzesEnabledPages;
+
+        if ( $wgEnableTriviaQuizzesExt && in_array( $wgTitle->getText(), $wgTriviaQuizzesEnabledPages ) ) {
```
Comment (Grunny) [inline extensions/wikia/TriviaQuizzes/TriviaQuizzesHooks.class.php]: Rather than using the global `$wgTitle`, it would be preferred to use `$out->getTitle()`. Also, `Title::getText` doesn't include the page's namespace, so the titles for both `Project:Foo` and `Foo` would both return `Foo`. As a result, I think using `Title::getPrefixedText` would be better here.

File: extensions/wikia/TriviaQuizzes/js/ext.TriviaQuizzes.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
+require(['jquery', 'content_types_consumption', 'wikia.trackingOptIn'], function ($, consumption, trackingOptIn) {
+	'use strict';
+	$(function() {
+		$('.WikiaRail').on('afterLoad.rail', function() {
+			// For now: if they are IE11, remove the entire module until we have better support
+			if($.browser.msie) {
+				$('#wikia-recent-activity').remove();
+			} else {
+				var cityId = window.wgCityId ? window.wgCityId : 0;
```
Comment (Grunny) [inline extensions/wikia/TriviaQuizzes/js/ext.TriviaQuizzes.js]: In newer versions of MediaWiki, it is preferred to access global JS vars via `mw.config.get`, i.e. `mw.config.get('wgCityId');`. You can require the `mw` module for this.

File: extensions/wikia/TriviaQuizzes/TriviaQuizzes.setup.php [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+$wgExtensionCredits['other'][] = [
+    'name' => 'Trivia Quizzes',
+    'description' => 'https://github.com/Wikia/app/tree/dev/extensions/wikia/TriviaQuizzes',
+];
+
+/**
+ * Resources Loader modules
+ */
+$wgResourceModules['ext.wikia.TriviaQuizzes'] = [
+    'remoteExtPath' => 'wikia/TriviaQuizzes',
```
Comment (Grunny) [inline extensions/wikia/TriviaQuizzes/TriviaQuizzes.setup.php]: Minor: Spaces -> tabs.

File: skins/oasis/modules/LatestActivityController.class.php [hunk 9]
```diff
@@ -58,6 +59,15 @@ public function executeIndex() {
...
 		$this->setVal( 'renderCommunityEntryPoint', !empty( $wgEnableCommunityPageExt ) );
 
+		// TODO https://wikia-inc.atlassian.net/browse/CAKE-4746
+		if ( $wgEnableTriviaQuizzesExt && in_array($wgTitle->getText(), $wgTriviaQuizzesEnabledPages) ) {
+            $this->setVal( 'renderTriviaQuizzes', $wgEnableTriviaQuizzesExt );
+            // TODO use this string instead when we have i18n support, for now, hard code the string for all languages
+            // $this->setVal( 'moduleHeader', wfMessage('trivia-quizzes-featured-quizzes-header')->escaped() );
+            $this->setVal( 'moduleHeader', 'Featured Quizzes');
+            $this->setVal( 'activityIcon', DesignSystemHelper::renderSvg( 'wds-icons-checkbox-small', 'wds-icon wds-icon-small wds-checkbox-icon' ) );
+        }
```
Comment (Grunny) [inline skins/oasis/modules/LatestActivityController.class.php]: Minor: Spaces -> tabs for all the added lines.

File: skins/oasis/modules/LatestActivityController.class.php [hunk 10]
```diff
@@ -58,6 +59,15 @@ public function executeIndex() {
 		$this->setVal( 'activityIcon', DesignSystemHelper::renderSvg( 'wds-icons-activity-small', 'wds-icon wds-icon-small wds-activity-icon' ) );
 		$this->setVal( 'renderCommunityEntryPoint', !empty( $wgEnableCommunityPageExt ) );
 
+		// TODO https://wikia-inc.atlassian.net/browse/CAKE-4746
+		if ( $wgEnableTriviaQuizzesExt && in_array($wgTitle->getText(), $wgTriviaQuizzesEnabledPages) ) {
```
Comment (Grunny) [inline skins/oasis/modules/LatestActivityController.class.php]: Similarly here as above, `$wgTitle->getText()` could be `$this->getContext()->getTitle()->getPrefixedText()`.

File: skins/oasis/modules/LatestActivityController.class.php [hunk 11]
```diff
@@ -58,6 +59,15 @@ public function executeIndex() {
...
+            $this->setVal( 'renderTriviaQuizzes', $wgEnableTriviaQuizzesExt );
+            // TODO use this string instead when we have i18n support, for now, hard code the string for all languages
+            // $this->setVal( 'moduleHeader', wfMessage('trivia-quizzes-featured-quizzes-header')->escaped() );
```
Comment (Grunny) [inline skins/oasis/modules/LatestActivityController.class.php]: Would it be worth just using this now? All other languages will fallback to the English string if no translations exist anyway, so right now it would be the same result as hardcoding and avoids needing to change it later, since once translations are added, they'll just work.

File: skins/oasis/modules/templates/LatestActivity_Index.php [hunk 12]
```diff
@@ -1,28 +1,33 @@
 <section class="rail-module activity-module" id="wikia-recent-activity">
 
-	<h2 class="has-icon"><?= $activityIcon ?><?= $moduleHeader ?></h2>
+    <h2 class="has-icon"><?= $activityIcon ?><?= $moduleHeader ?></h2>
```
Comment (Grunny) [inline skins/oasis/modules/templates/LatestActivity_Index.php]: Minor: The whole file should be using tabs per our PHP conventions.

**Review by Grunny — CHANGES_REQUESTED** [review]

File: skins/oasis/modules/LatestActivityController.class.php (hunk 11 above)
Comment (jamieshepherd) [reply skins/oasis/modules/LatestActivityController.class.php]: Good to know, thanks!

File: extensions/wikia/TriviaQuizzes/js/ext.TriviaQuizzes.js [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+				var dbName = window.wgDBname ? window.wgDBname : null;
+				var user = window.wgUserName ? { id: 123, isGDPRApproved: trackingOptIn.isOptedIn() } : null;
```
Comment (jamieshepherd) [inline extensions/wikia/TriviaQuizzes/js/ext.TriviaQuizzes.js]: TODO: Add user ID here

File: extensions/wikia/TriviaQuizzes/TriviaQuizzesHooks.class.php (hunk 5 above)
Comment (vforge) [reply extensions/wikia/TriviaQuizzes/TriviaQuizzesHooks.class.php]: Fixed in d02b0da8788

File: extensions/wikia/TriviaQuizzes/TriviaQuizzes.setup.php (hunk 8 above)
Comment (vforge) [reply extensions/wikia/TriviaQuizzes/TriviaQuizzes.setup.php]: Fixed in d02b0da8788

File: skins/oasis/modules/LatestActivityController.class.php (hunk 10 above)
Comment (vforge) [reply skins/oasis/modules/LatestActivityController.class.php]: Fixed in 93c4c1a7be7

File: skins/oasis/modules/templates/LatestActivity_Index.php (hunk 12 above)
Comment (vforge) [reply skins/oasis/modules/templates/LatestActivity_Index.php]: Fixed in 93c4c1a7be7

File: skins/oasis/modules/LatestActivityController.class.php (hunk 11 above)
Comment (vforge) [reply skins/oasis/modules/LatestActivityController.class.php]: Fixed in 93c4c1a7be7

File: extensions/wikia/TriviaQuizzes/TriviaQuizzesHooks.class.php (hunk 6 above)
Comment (vforge) [reply extensions/wikia/TriviaQuizzes/TriviaQuizzesHooks.class.php]: Fixed in de8ccb30cf8

File: extensions/wikia/TriviaQuizzes/js/ext.TriviaQuizzes.js (hunk 7 above)
Comment (vforge) [reply extensions/wikia/TriviaQuizzes/js/ext.TriviaQuizzes.js]: Fixed in 88f7f44be9d

File: extensions/wikia/TriviaQuizzes/js/ext.TriviaQuizzes.js (hunk 13 above)
Comment (vforge) [reply extensions/wikia/TriviaQuizzes/js/ext.TriviaQuizzes.js]: Fixed in 88f7f44be9d

**Comment by vforge** [comment]
@Grunny I think I fixed all the issues mentioned.
One thing that I want to change before we merge this is that we'll use newer version of the library (we're working on this right now), but this will be the only change. The rest will stay as-is.

File: extensions/wikia/TriviaQuizzes/js/ext.TriviaQuizzes.js [hunk 14]
```diff
@@ -0,0 +1,25 @@
+require(['jquery', 'content_types_consumption', 'wikia.trackingOptIn', 'mw'], function ($, consumption, trackingOptIn) {
```
Comment (Grunny) [inline extensions/wikia/TriviaQuizzes/js/ext.TriviaQuizzes.js]: `mw` should probably be on the list of method params here.

```suggestion
require(['jquery', 'content_types_consumption', 'wikia.trackingOptIn', 'mw'], function ($, consumption, trackingOptIn, mw) {
```

File: extensions/wikia/TriviaQuizzes/TriviaQuizzesHooks.class.php [hunk 15]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+class TriviaQuizzesHooks extends WikiaController {
+	public static function onBeforePageDisplay( OutputPage $out ) {
+		global $wgTitle, $wgEnableTriviaQuizzesExt, $wgTriviaQuizzesEnabledPages;
+
+		if ( $wgEnableTriviaQuizzesExt && in_array( $wgTitle->getPrefixedText(), $wgTriviaQuizzesEnabledPages ) ) {
```
Comment (Grunny) [inline extensions/wikia/TriviaQuizzes/TriviaQuizzesHooks.class.php]:
```suggestion
		if ( $wgEnableTriviaQuizzesExt && in_array( $out->getTitle()->getPrefixedText(), $wgTriviaQuizzesEnabledPages ) ) {
```

And as a result remove `$wgTitle` global declaration above.

File: skins/oasis/modules/LatestActivityController.class.php [hunk 16]
```diff
@@ -58,6 +60,14 @@ public function executeIndex() {
 		$this->setVal( 'activityIcon', DesignSystemHelper::renderSvg( 'wds-icons-activity-small', 'wds-icon wds-icon-small wds-activity-icon' ) );
 		$this->setVal( 'renderCommunityEntryPoint', !empty( $wgEnableCommunityPageExt ) );
 
+		// TODO: https://wikia-inc.atlassian.net/browse/CAKE-4746
+		$currentPageTitle = $this->getContext()->getTitle()->getPrefixedText();
+		if ( $wgEnableTriviaQuizzesExt && in_array($currentPageTitle, $wgTriviaQuizzesEnabledPages) ) {
+				$this->setVal( 'renderTriviaQuizzes', $wgEnableTriviaQuizzesExt );
+				$this->setVal( 'moduleHeader', wfMessage('trivia-quizzes-featured-quizzes-header')->escaped() );
+				$this->setVal( 'activityIcon', DesignSystemHelper::renderSvg( 'wds-icons-checkbox-small', 'wds-icon wds-icon-small wds-checkbox-icon' ) );
```
Comment (Grunny) [inline skins/oasis/modules/LatestActivityController.class.php]: Minor: remove an indentation level on these three.

File: skins/oasis/modules/LatestActivityController.class.php [hunk 17]
```diff
@@ -58,6 +60,14 @@ public function executeIndex() {
 		$this->setVal( 'activityIcon', DesignSystemHelper::renderSvg( 'wds-icons-activity-small', 'wds-icon wds-icon-small wds-activity-icon' ) );
 		$this->setVal( 'renderCommunityEntryPoint', !empty( $wgEnableCommunityPageExt ) );
 
+		// TODO: https://wikia-inc.atlassian.net/browse/CAKE-4746
+		$currentPageTitle = $this->getContext()->getTitle()->getPrefixedText();
+		if ( $wgEnableTriviaQuizzesExt && in_array($currentPageTitle, $wgTriviaQuizzesEnabledPages) ) {
```
Comment (Grunny) [inline skins/oasis/modules/LatestActivityController.class.php]:
```suggestion
		if ( $wgEnableTriviaQuizzesExt && in_array( $currentPageTitle, $wgTriviaQuizzesEnabledPages ) ) {
```
Just adding some extra spaces per our coding conventions.

**Review by Grunny — CHANGES_REQUESTED** [review]
@vforge Looks good, just made some final small comments, suggestions.

File: skins/oasis/modules/LatestActivityController.class.php (hunk 17 above)
Comment (vforge) [reply skins/oasis/modules/LatestActivityController.class.php]: done in b28949d

File: skins/oasis/modules/LatestActivityController.class.php (hunk 16 above)
Comment (vforge) [reply skins/oasis/modules/LatestActivityController.class.php]: done in b28949d

File: extensions/wikia/TriviaQuizzes/TriviaQuizzesHooks.class.php (hunk 15 above)
Comment (vforge) [reply extensions/wikia/TriviaQuizzes/TriviaQuizzesHooks.class.php]: done in b28949d

**Comment by vforge** [comment]
@Grunny cheers

**Review by Grunny — APPROVED** [review]

**Comment by vforge** [comment]
Actually I'm going to merge it right now and then bump the lib in separate PR - if that's fine with you, @Grunny

**Comment by Grunny** [comment]
Yep, looks good to merge to me. :)
